QccPack-0.61-1/src/libQccPackWAV/utils/man1/make_noise_image.1
.TH MAKE_NOISE_IMAGE 1 "QCCPACK" ""
.SH NAME
make_noise_image \- make a noise image for perceptual testing
.SH SYNOPSIS
.B make_noise_image
.RB "[\|" \-f
.IR  wavelet "\|]"
.RB "[\|" \-b
.IR  boundary "\|]"
.RB "[\|" \-nl
.IR levels "\|]"
.RB "[\|" \-s 
.IR subband "\|]"
.RB "[\|" \-w
.IR image_width "\|]"
.RB "[\|" \-h
.IR image_height "\|]"
.RB "[\|" \-b
.IR background "\|]"
.RB "[\|" \-n
.IR noise_variance "\|]"
.I imgfile
.SH OPTIONS
.TP
.BI \-w " wavelet"
String.
Name of the wavelet file
.RB ( FBK
or
.B LFT
format); default = "CohenDaubechiesFeauveau.9-7.lft"
.TP
.BI \-b " boundary"
String. Type of boundary extension; default = "symmetric".
.TP
.BI \-nl " levels"
Integer.  Number of levels of dyadic decomposition.  Default is 1.
.TP
.BI \-s " subband"
Integer. Number of the subband in which to make the noise image. Default is
0 (the low-frequency, baseband subband).
.TP
.BI \-w " image_width"
Integer. Width (number of columns) of the output image
.IR imgfile .
Default is 512.
.TP
.BI \-h " image_height"
Integer. Height (number of rows) of the output image,
.IR imgfile .
Default is 512.
.TP
.BI \-b " background"
Float. Background level of image; i.e., the mean added to the image after
subband reconstruction.  Default value is 127.
.TP
.BI \-n " noise_variance"
Float. Variance of the noise. Default value is 1.0.
.SH DESCRIPTION
.LP
.B make_noise_image
creates an image with noise to be used in perceptual tests.
.B make_noise_image
creates an image consisting of a solid background (the value of which
is given by the
.B \-b
option) with a square region of noise added to the center.

More precisely,
.B make_noise_image
creates a subband pyramid of depth specified by
.I levels
(option
.BR \-nl )
and initializes it to all zeros.
Then, noise of variance
.I noise_variance
(option
.BR \-n )
is added in a square in the center of the subband specified by
option
.BR \-s .
.B make_noise_image
then performs a 2D separable wavelet reconstruction
of the subband pyramid (see
.BR imgidwt (1)).
Finally, mean value
.I background
(option
.BR \-b )
is added to the entire image, and the result is output to
.BR imgfile ,
a PGM image.
.LP
The noise added to the image is zero-mean, uniform, and has variance
.IR noise_variance ,
i.e., the range of the noise  is
.I -c
to
.I c
where
.IR "c = " sqrt(3 " * noise_variance" ).
.LP
.B make_noise_image
is intended to be used in the determining of perceptual weightings as
outlined in Jayant et al. and in Hontcsh et al.
.B make_noise_image
can be used to output images at a number of noise levels, with
a number of background intensities, and with noise in different subbands
of the pyramidal subband decomposition.  An observer 
can then determine, visually,
which level of noise gives 
.I "just-noticeable distortion"
(JND) in each subband for each background intensity.
After a series of such perceptual experiments, the frequency and brightness
sensitivity as a function of the subband number (i.e., frequency)
may be determined.
.LP
Note: If 
.I levels
is zero, no subband reconstruction is performed.  In this case, noise is
added to the baseband subband (i.e., the only band), the mean is added, and the result is output directly.
.SH "SEE ALSO"
.BR imgidwt (1),
.BR QccPackWAV (3),
.BR QccPackIMG (3),
.BR QccPack (3)

N. Jayant, J. Johnston, and R. Safranek,
"Signal Compression Based on Models of Human Perception,"
.IR "Proceedings of the IEEE" ,
vol. 81, pp. 1385-1422, October 1993.

I. Hontsch, L. J. Karam, and R. J. Safranek,
"A Perceptually Tuned Embedded Zerotree Image Coder,"
.IR "Proceedings of the IEEE International Conference on Image Processing" ,
vol. 1, pp. 41-44, October 1997.

M. Antonini, M. Barlaud, P. Mathieu, and I. Daubechies,
"Image Coding Using Wavelet Transform,"
.IR "IEEE Transactions on Image Processing" ,
vol. 1, pp. 205-220, April 1992.

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackWAV/utils/make_noise_image.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#include "make_noise_image.h"

#define SQUARE_START 0.25
#define SQUARE_WIDTH 0.5

#define USG_STRING "[-w %s:wavelet] [-b %s:boundary] [-nl %d:levels] [-s %d:subband] [-w %d:image_width] [-h %d:image_height] [-bg %f:background] [-n %f:noise_variance] %s:imgfile"

QccWAVWavelet Wavelet;
QccString WaveletFilename = QCCWAVWAVELET_DEFAULT_WAVELET;
QccString Boundary = "symmetric";

int NumLevels = 1;
int Subband = 0;
int ImageWidth = 512;
int ImageHeight = 512;
float Background = 127;
float NoiseVariance = 1.0;

int NumSubbands;

QccWAVSubbandPyramid SubbandPyramid;
QccIMGImage OutputImage;

int main(int argc, char *argv[])
{
  int row, col;
  
  QccInit(argc, argv);
  
  QccWAVWaveletInitialize(&Wavelet);
  QccWAVSubbandPyramidInitialize(&SubbandPyramid);
  QccIMGImageInitialize(&OutputImage);

  if (QccParseParameters(argc, argv,
                         USG_STRING,
                         WaveletFilename,
                         Boundary,
                         &NumLevels,
                         &Subband,
                         &ImageWidth,
                         &ImageHeight,
                         &Background,
                         &NoiseVariance,
                         OutputImage.filename))
    QccErrorExit();
  
  if (QccWAVWaveletCreate(&Wavelet, WaveletFilename, Boundary))
    {
      QccErrorAddMessage("%s: Error calling QccWAVWaveletCreate()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (NumLevels < 0)
    {
      QccErrorAddMessage("%s: Number of levels must be nonnegative",
                         argv[0]);
      QccErrorExit();
    }
  NumSubbands = QccWAVSubbandPyramidNumLevelsToNumSubbands(NumLevels);
  if ((ImageWidth < 1) || (ImageHeight < 1))
    {
      QccErrorAddMessage("%s: Invalid image size (%d x %d)",
                         argv[0], ImageWidth, ImageHeight);
      QccErrorExit();
    }
  if ((ImageWidth != ((int)(ImageWidth >> NumLevels) << NumLevels)) ||
      (ImageHeight != ((int)(ImageHeight >> NumLevels) << NumLevels)))
    {
      QccErrorAddMessage("%s: Image size (%d x %d) is not a multiple of %d (as needed for a decomposition of %d levels)",
                         argv[0],
                         ImageWidth, ImageHeight,
                         (1 << NumLevels), NumLevels);
      QccErrorExit();
    }
  if ((Subband < 0) || (Subband >= NumSubbands))
    {
      QccErrorAddMessage("%s: Requested subband %d is invalid for a %d-level decomposition)",
                         argv[0], Subband, NumLevels);
      QccErrorExit();
    }
  if (NoiseVariance < 0.0)
    {
      QccErrorAddMessage("%s: Invalid noise variance (%f)",
                         argv[0], NoiseVariance);
      QccErrorExit();
    }
  
  SubbandPyramid.num_rows = ImageHeight;
  SubbandPyramid.num_cols = ImageWidth;
  SubbandPyramid.num_levels = NumLevels;
  if (QccWAVSubbandPyramidAlloc(&SubbandPyramid))
    {
      QccErrorAddMessage("%s: Error calling QccWAVSubbandPyramidAlloc()",
                         argv[0]);
      QccErrorExit();
    }
  for (row = 0; row < ImageHeight; row++)
    for (col = 0; col < ImageWidth; col++)
      SubbandPyramid.matrix[row][col] = 0.0;
  
  if (QccWAVSubbandPyramidAddNoiseSquare(&SubbandPyramid,
                                         Subband,
                                         SQUARE_START,
                                         SQUARE_WIDTH,
                                         (double)NoiseVariance))
    {
      QccErrorAddMessage("%s: Error calling QccWAVSubbandPyramidAddNoiseSquare()",
                         argv[0]);
      QccErrorExit();
    }
  if (NumLevels)
    if (QccWAVSubbandPyramidInverseDWT(&SubbandPyramid,
                                       &Wavelet))
      {
        QccErrorAddMessage("%s: Error calling QccWAVSubbandPyramidInverseDWT()",
                           argv[0]);
        QccErrorExit();
      }
  
  if (QccWAVSubbandPyramidAddMean(&SubbandPyramid,
                                  (double)Background))
    {
      QccErrorAddMessage("%s: Error calling QccWAVSubbandPyramidAddMean()",
                         argv[0]);
      QccErrorExit();
    }

  OutputImage.Y.num_rows = SubbandPyramid.num_rows;
  OutputImage.Y.num_cols = SubbandPyramid.num_cols;
  OutputImage.Y.image = SubbandPyramid.matrix;

  if (QccIMGImageSetMaxMin(&OutputImage))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageSetMaxMin()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (QccIMGImageWrite(&OutputImage))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageWrite()",
                         argv[0]);
      QccErrorExit();
    }
  
  QccWAVWaveletFree(&Wavelet);
  QccIMGImageFree(&OutputImage);

  QccExit;
}
